libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m

```diff
@@ -297,66 +297,75 @@ - (BOOL)authenticateWithCompletion:(SFUserAccountManagerSuccessCallbackBlock)com
 }
 
 - (BOOL)loginWithJwtToken:(NSString *)jwtToken completion:(SFUserAccountManagerSuccessCallbackBlock)completionBlock failure:(SFUserAccountManagerFailureCallbackBlock)failureBlock {
-    NSAssert(self.useLegacyAuthenticationManager==false, kSFIncompatibleAuthError);
+    NSAssert(self.useLegacyAuthenticationManager == false, kSFIncompatibleAuthError);
     NSAssert(jwtToken.length > 0, @"JWT token value required.");
     SFOAuthCredentials *credentials = [self newClientCredentials];
     credentials.jwt = jwtToken;
     return [self authenticateWithCompletion:completionBlock failure:failureBlock credentials:credentials];
 }
 
 - (void)logout {
-    NSAssert(self.useLegacyAuthenticationManager==false, kSFIncompatibleAuthError);
+    NSAssert(self.useLegacyAuthenticationManager == false, kSFIncompatibleAuthError);
     [self logoutUser:[SFUserAccountManager sharedInstance].currentUser];
 }
 
 - (void)logoutUser:(SFUserAccount *)user {
-    NSAssert(self.useLegacyAuthenticationManager==false, kSFIncompatibleAuthError);
+    NSAssert(self.useLegacyAuthenticationManager == false, kSFIncompatibleAuthError);
+
     // No-op, if the user is not valid.
     if (user == nil) {
-        [SFSDKCoreLogger i:[self class] format:@"logoutUser: user is nil.  No action taken."];
+        [SFSDKCoreLogger i:[self class] format:@"logoutUser: user is nil. No action taken."];
         return;
     }
-
     BOOL loggingOutTransitionSucceeded = [user transitionToLoginState:SFUserAccountLoginStateLoggingOut];
     if (!loggingOutTransitionSucceeded) {
+
         // SFUserAccount already logs the transition failure.
         return;
     }
-    
-    BOOL isCurrentUser = [user isEqual:self.currentUser];
     [SFSDKCoreLogger d:[self class] format:@"Logging out user '%@'.", user.userName];
     NSDictionary *userInfo = @{ kSFNotificationUserInfoAccountKey : user };
     [[NSNotificationCenter defaultCenter]  postNotificationName:kSFNotificationUserWillLogout
                                                         object:self
                                                       userInfo:userInfo];
+    if ([SFPushNotificationManager sharedInstance].deviceSalesforceId) {
+        __weak typeof(self) weakSelf = self;
+        [[SFPushNotificationManager sharedInstance] unregisterSalesforceNotificationsWithCompletionBlock:user completionBlock:^void() {
+            __strong typeof(weakSelf) strongSelf = weakSelf;
+            [strongSelf postPushUnregistration:userInfo user:user];
+        }];
+    } else {
+        [self postPushUnregistration:userInfo user:user];
+    }
+}
+
+- (void)postPushUnregistration:(NSDictionary *)userInfo user:(SFUserAccount *)user {
     SFSDKOAuthClient *client = [self fetchOAuthClient:user.credentials completion:nil failure:nil];
     [self deleteAccountForUser:user error:nil];
     [client cancelAuthentication:NO];
     [client revokeCredentials];
-    if ([SFPushNotificationManager sharedInstance].deviceSalesforceId) {
```

Comment:
I'm not sure if/how this was working before since the unregistration call is clearly being made after `revokeCredentials`.

Reply:
That's the 6.0 regression.
Here is the 5.3 code: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/v5.3.1/libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFAuthenticationManager.m#L446